Source/DafnyCore/AST/MemberDecls.cs

```diff
@@ -466,6 +467,34 @@ public class Method : MemberDecl, TypeParameter.ParentType, IMethodCodeContext {
   public override bool IsOverrideThatAddsBody => base.IsOverrideThatAddsBody && Body != null;
   private static BlockStmt emptyBody = new BlockStmt(Token.NoToken, Token.NoToken, new List<Statement>());
 
+  public override IEnumerable<AssumptionDescription> Assumptions {
+    get {
+      foreach (var a in base.Assumptions) {
+        yield return a;
+      }
+
+      if (Children.Any(n => n is AssumeStmt)) {
+        yield return AssumptionDescription.AssumeInBody;
+      }
+
+      if (Body is null && Ens.Count > 0) {
+        yield return AssumptionDescription.NoBody(IsGhost);
+      }
+
+      if (Attributes.Contains(Attributes, "extern") && Ens.Count > 0) {
+        yield return AssumptionDescription.ExternWithEnsures;
+      }
+
+      if (Attributes.Contains(Attributes, "extern") && Req.Count > 0) {
+        yield return AssumptionDescription.ExternWithRequires;
+      }
+
+      if (AllowsNontermination) {
```

Comment:
Should this show up in the report even if the method has no ensures clauses? Many programs can potentially run forever by design, and in Dafny you need to use `decreases *` to write those programs.

```dafny
// This method looks fine. Why would you report it?
method LoopForever() decreases * {
  while(true) decreases * {
    DoUsefulThing();
  }
}
```

```dafny
// The ensures clause is misleading, because you can't actually make use of it since a call to PointlessEnsures never finishes. I think this should be a warning.
method PointlessEnsures() 
  decreases * 
  ensures 1 == 2 {
  PointlessEnsures();
}
```

Reply:
I'm somewhat undecided about this, myself. Clearly a method with `decreases *` _and_ an `ensures` clause is an issue. Even without one, though, I think they're worth calling out, though I don't feel super strongly about that. They are necessary for certain programs, but they should also be rare. The key situation I can see occurring is that you put in `decreases *` at first because termination is hard to prove, but with the intention of removing it, and then forget to do so.

Reply:
It could also be that you are writing an interactive program for an arbitrary number of human commands before you exit the loop. In this case, there is no choice but use `decreases *`. However, you still want to ensure postconditions.
It would make sense down the road to generate a warning if we can prove statically that a method marked with `decreases *` can never terminate and has postconditions. For example, if the main execution path goes through a while loop and there is no break or return statement. But I think that would be a separate feature.

Reply:
I think I'm now leaning toward not even including it in the auditor, but I'm not 100% convinced. Yes, there are definitely cases where you can't avoid using `decreases *`, so I don't think it should be a default warning. But the case where you use it as a way to temporarily avoid needing to prove termination and then forget to remove it does seem like a legitimate mistake that could undermine the guarantees that you believe you've achieved.

Reply:
That's a legitimate concern. Let me alleviate it.
If in a random method, if you use decreases * to avoid proving termination, you'll actually introduce more problems than you'll solve. This is because Dafny will require you to put decreases * in the method signature as well and in all the calling methods.
This is never worth the hassle in development, so to avoid proving termination, one would usually write the assume of the termination (or assume false), which the auditor would check.

However, as part of a comprehensive report, it would still be worth knowing if the application is suppose to terminate or not. So reporting the name of all methods that use decreases * makes sense, so that someone auditing the code would verify that there are only a very small amount of them.
You could also emit a warning if a method name with decreases * does not contain Loop(), Main() or Run() for example.
But for the current auditor, I'm in favor of not taking into account decreases * because of their constraints that prevents them from being used as scaffolding.